### PR TITLE
Fix null type argument inference with T?

### LINKED ISSUE #56788: CS8620, CS8634 mis-firing on extension method invocation with implied generic type args
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/CS8620-CS8634-mis-firing-on-extension-m/1534336)._

---
[severity:It bothers me. A fix would be nice] [regression] [worked-in:16.11]

```
git clone https://github.com/microsoft/vs-mef.git
cd vs-mef
git checkout 67750f0c7b2ada5880ad2f937560f4051af333fb
.\Microsoft.VisualStudio.Composition.sln
```

Open `src\Microsoft.VisualStudio.Composition\ExportProvider.cs` within VS.

**Expected**

No errors or warnings in the error list.

**Actual**

Two warnings appear in the error list.

Severity	| Code	| Description	| Project	| File	| Line	| Suppression State
--|--|--|--|--|--|--
Warning	| CS8634	| The type 'System.Collections.Generic.Dictionary<Microsoft.VisualStudio.Composition.Reflection.TypeRef, Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker>?' cannot be used as type parameter 'TValue' in the generic type or method 'Utilities.GetValueOrDefault<TKey, TValue>(IReadOnlyDictionary<TKey, TValue>, TKey, TValue?)'. Nullability of type argument 'System.Collections.Generic.Dictionary<Microsoft.VisualStudio.Composition.Reflection.TypeRef, Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker>?' doesn't match 'class' constraint.	|Microsoft.VisualStudio.Composition (net472), Microsoft.VisualStudio.Composition (netcoreapp3.1), Microsoft.VisualStudio.Composition (netstandard2.0)	| C:\git\MEF\src\Microsoft.VisualStudio.Composition\ExportProvider.cs	|994	| Active
Warning	| CS8620	| Argument of type 'ImmutableDictionary<string, Dictionary<TypeRef, ExportProvider.PartLifecycleTracker>>' cannot be used for parameter 'dictionary' of type 'IReadOnlyDictionary<string, Dictionary<TypeRef, ExportProvider.PartLifecycleTracker>?>' in 'Dictionary<TypeRef, PartLifecycleTracker>? Utilities.GetValueOrDefault<string, Dictionary<TypeRef, PartLifecycleTracker>?>(IReadOnlyDictionary<string, Dictionary<TypeRef, PartLifecycleTracker>?> dictionary, string key, Dictionary<TypeRef, PartLifecycleTracker>? defaultValue)' due to differences in the nullability of reference types.	|Microsoft.VisualStudio.Composition (net472), Microsoft.VisualStudio.Composition (netcoreapp3.1), Microsoft.VisualStudio.Composition (netstandard2.0)	| C:\git\MEF\src\Microsoft.VisualStudio.Composition\ExportProvider.cs	|994	| Active

When I explicitly specify the type arguments, the warnings go away:

```diff
-var sharingBoundary = this.sharedInstantiatedParts.GetValueOrDefault(sharingBoundaryName, null);
+var sharingBoundary = this.sharedInstantiatedParts.GetValueOrDefault<string, Dictionary<TypeRef, PartLifecycleTracker>>(sharingBoundaryName, null);
```

But then the language service suggests I remove the type arguments to "simplify" the line, but removing them brings those two warnings back.

---
#### Feedback Bot on 9/22/2021, 03:37 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #56788

When a type parameter is annotated, you can always read null from it and/or write null to it. We only want `null` to contribute its nullability if there is a possibility that the type parameter could be substituted for a non-nullable reference type.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -73094,6 +73094,100 @@ void M1()
...
+        public void TypeArgumentInference_09()
+        {
+            var source = @"
+class C
+{
+    T M<T>(T t1, T? t2) => t2 ?? t1;
+
+    void M1()
+    {
+        var result1 = M(""a"", default);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:73107]: Consider combining with lambda: `T MT<T>(T t1, Func<T?> t2)` called with `("a", () => default)`

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1). Just a minor test suggestion

**Comment by jcouv** [comment]
@RikkiGibson FYI, I retargeted the PR to 17.0 branch. Hope that's okay

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:73107]: Done

**Comment by jcouv** [comment]
@cston is probably a good second reviewer since he'd reviewed https://github.com/dotnet/roslyn/pull/56022 (fixing type inference to account for `null` literals)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -73094,6 +73094,124 @@ void M1()
                 );
         }
 
+        [Fact]
+        public void TypeArgumentInference_09()
+        {
+            var source = @"
+class C
+{
+    T M<T>(T t1, T? t2) => t2 ?? t1;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:73103]: >T M<T>(T t1, T? t2) => t2 ?? t1; [](http://example.com/codeflow?start=4&length=32)

Consider testing with `T!` and `T~` as well.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:73103]: Adding the test, but not intending to modify the implementation for this scenario unless it feels really bad/unusable. Feels tangentially related to #56518 (what should happen to type arguments when type parameters are oblivious).

I think tests for not-annotated type parameters are already present. For example, `TypeArgumentInference_02`.

**Review by cston — APPROVED** [review]

**Pull request merged**